File: public/datasets/bls-qcew/derive-smaller.py
import csv
import os
import sys

INPUT = os.path.join(os.path.dirname(__file__), "bls_qcew_county.csv")

DERIVED = [
    {
        "file": "bls_qcew_county_all_industries.csv",
        "desc": "'All Industries' (code=00) for all years",
        "filter": lambda r: r["industry_code"] == "00",
    },
    {
        "file": "bls_qcew_county_all_industries_2024.csv",
        "desc": "'All Industries' (code=00) for 2024 only",
        "filter": lambda r: r["industry_code"] == "00" and r["year"] == "2024",
    },
    {
        "file": "bls_qcew_county_manufacturing.csv",
        "desc": "Manufacturing (code=31-33) for all years",
        "filter": lambda r: r["industry_code"] == "31-33",
    },
    {
        "file": "bls_qcew_county_2024.csv",
        "desc": "All industry codes for 2024 only",
        "filter": lambda r: r["year"] == "2024",
    },
]

def main():
    print("Reading source CSV...")
    rows = []
    with open(INPUT, "r") as f:
        reader = csv.DictReader(f)
        for row in reader:
            # Normalise whitespace in industry_name
            row["industry_name"] = " ".join(row["industry_name"].split())
            rows.append(row)
    print(f"  Read {len(rows)} rows\n")

    base_dir = os.path.dirname(__file__)

    for spec in DERIVED:
        filtered = [r for r in rows if spec["filter"](r)]
        out_path = os.path.join(base_dir, spec["file"])
        with open(out_path, "w", newline="") as f:
            writer = csv.DictWriter(f, fieldnames=filtered[0].keys())
            writer.writeheader()
            writer.writerows(filtered)
        size_kb = os.path.getsize(out_path) / 1024
        print(
            f"  ✅ {spec['file']}  "
            f"({len(filtered):,} rows, {size_kb:.0f} KB)  "
            f"{spec['desc']}"
        )

    print("\nAll 4 derived files created.")


if __name__ == "__main__":
    main()

File: public/datasets/bls-qcew/derive-smaller.mjs
import { createReadStream, createWriteStream, statSync } from "fs";
import { parse } from "csv-parse";
import { stringify } from "csv-stringify/sync";

const INPUT = new URL("bls_qcew_county.csv", import.meta.url).pathname;

// Output definitions: [filename, description, filterFn]
const DERIVED = [
  {
    file: "bls_qcew_county_all_industries.csv",
    desc: "'All Industries' (code=00) for all years, all counties",
    filter: (r) => r.industry_code === "00",
  },
  {
    file: "bls_qcew_county_all_industries_2024.csv",
    desc: "'All Industries' (code=00) for 2024 only",
    filter: (r) => r.industry_code === "00" && r.year === "2024",
  },
  {
    file: "bls_qcew_county_manufacturing.csv",
    desc: "Manufacturing (code=31-33) for all years, all counties",
    filter: (r) => r.industry_code === "31-33",
  },
  {
    file: "bls_qcew_county_2024.csv",
    desc: "All industry codes for 2024 only",
    filter: (r) => r.year === "2024",
  },
];

async function main() {
  console.log("Reading source CSV...");
  const rows = [];
  const parser = createReadStream(INPUT).pipe(
    parse({ columns: true, relax_quotes: true, relax_column_count: true })
  );
  for await (const row of parser) {
    // Normalise whitespace in industry_name
    if (row.industry_name) {
      row.industry_name = row.industry_name.replace(/\s+/g, " ").trim();
    }
    rows.push(row);
  }
  console.log(`  Read ${rows.length} rows\n`);

  const STATS = [];

  for (const { file, desc, filter } of DERIVED) {
    const filtered = rows.filter(filter);
    const csv = stringify(filtered, { header: true });
    const outPath = new URL(file, import.meta.url).pathname;
    createWriteStream(outPath).end(csv);
    const stats = statSync(outPath);
    STATS.push({ file, desc, rows: filtered.length, size: stats.size });
    console.log(
      `  ✅ ${file}  (${filtered.length.toLocaleString()} rows, ${(stats.size / 1024).toFixed(0)} KB)  ${desc}`
    );
  }

  console.log(`\nAll 4 derived files created.`);
}

main().catch((err) => {
  console.error("Fatal:", err);
  process.exit(1);
});